File: leetcode/17.电话号码的数字组合/main.go
package main

import "fmt"

// global declaration
var combinations []string

var phone = map[string]string{
	"2": "abc",
	"3": "def",
	"4": "ghi",
	"5": "jkl",
	"6": "mno",
	"7": "pqrs",
	"8": "tuv",
	"9": "wxyz",
}

// letterCombinations is entry function.
func letterCombinations(digits string) []string {
	if digits == "" {
		return []string{}
	}

	combinations = []string{}
	backTrack("", digits, 0)

	return combinations
}

/*
* backTrack
* combination: the all concatenated letters at the moment.
* digits: original digits
* index: the index of digits string which is based from 0.
 */
func backTrack(combination string, digits string, index int) {
	// 当index和原始数字字母表的长度相同时, 不再回溯
	if index == len(digits) {
		combinations = append(combinations, combination)
	} else {
		// 获取对应的数字
		digit := string(digits[index])
		// 获取对应数字的所有字母
		letters := phone[digit]
		// 获取对应数字的所有字母的长度
		lettersCount := len(letters)

		// 取当前数字对应的字母表的第一个字母, 进行回溯
		for i := 0; i < lettersCount; i++ {
			backTrack(combination+string(letters[i]), digits, index+1)
		}
	}
}

func main() {
	fmt.Println(letterCombinations("24"))
}

File: leetcode/17.电话号码的数字组合/main.py
from collections import deque
from typing import List


class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []

        phone = {
            "2": ["a", "b", "c"],
            "3": ["d", "e", "f"],
            "4": ["g", "h", "i"],
            "5": ["j", "k", "l"],
            "6": ["m", "n", "o"],
            "7": ["p", "q", "r", "s"],
            "8": ["t", "u", "v"],
            "9": ["w", "x", "y", "z"],
        }

        def back_track(combination: str, rest_digit: List[str]):
            # 执行结果收集
            if rest_digit == "":
                result.append(combination)
            else:
                # 执行字符串的拼接
                for letter in phone[rest_digit[0]]:
                    back_track(combination + letter, rest_digit[1:])

        result = []
        back_track("", digits)

        return result

    def letterCombinations1(self, digits: str) -> List[str]:
        if not digits:
            return []
        phone = ["abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]
        queue = deque([""])  # 初始化队列
        for digit in digits:
            for _ in range(len(queue)):
                tmp = queue.popleft()
                for letter in phone[ord(digit) - 50]:  # 这里我们不使用 int() 转换字符串，使用ASCII码
                    queue.append(tmp + letter)

        # Convert deque to list.
        return list(queue)


if __name__ == "__main__":
    print(Solution().letterCombinations1("24"))
